render/pixman/renderer.c

```diff
@@ -0,0 +1,379 @@
+
+	pixman_image_composite32(PIXMAN_OP_SRC, fill, NULL, buffer->image, 0, 0, 0,
+			0, 0, 0, renderer->width, renderer->height);
+
+	pixman_image_unref(fill);
+}
+
+static void pixman_scissor(struct wlr_renderer *wlr_renderer,
+		struct wlr_box *box) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	struct wlr_pixman_buffer *buffer = renderer->current_buffer;
+
+	if (box != NULL) {
+		struct pixman_region32 region = {0};
+		pixman_region32_init_rect(&region, box->x, box->y, box->width,
+				box->height);
+		pixman_image_set_clip_region32(buffer->image, &region);
+		pixman_region32_fini(&region);
+	} else {
+		pixman_image_set_clip_region32(buffer->image, NULL);
+	}
+}
+
+static void matrix_to_pixman_transform(struct pixman_transform *transform,
+		const float mat[static 9]) {
+	struct pixman_f_transform ftr;
+	ftr.m[0][0] = mat[0];
+	ftr.m[0][1] = mat[1];
+	ftr.m[0][2] = mat[2];
+	ftr.m[1][0] = mat[3];
+	ftr.m[1][1] = mat[4];
+	ftr.m[1][2] = mat[5];
+	ftr.m[2][0] = mat[6];
+	ftr.m[2][1] = mat[7];
+	ftr.m[2][2] = mat[8];
+
+	pixman_transform_from_pixman_f_transform(transform, &ftr);
+}
+
+static bool pixman_render_subtexture_with_matrix(
+		struct wlr_renderer *wlr_renderer, struct wlr_texture *wlr_texture,
+		const struct wlr_fbox *fbox, const float matrix[static 9],
+		float alpha) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+	struct wlr_pixman_buffer *buffer = renderer->current_buffer;
+
+	// TODO: don't create a mask if alpha == 1.0
+	struct pixman_color mask_colour = {0};
+	mask_colour.alpha = 0xFFFF * alpha;
+	pixman_image_t *mask = pixman_image_create_solid_fill(&mask_colour);
+
+	float m[9];
+	memcpy(m, matrix, sizeof(m));
+	wlr_matrix_scale(m, 1.0 / fbox->width, 1.0 / fbox->height);
+
+	struct pixman_transform transform = {0};
+	matrix_to_pixman_transform(&transform, m);
+	pixman_transform_invert(&transform, &transform);
+
+	pixman_image_set_transform(texture->image, &transform);
+
+	// TODO clip properly with src_x and src_y
+	pixman_image_composite32(PIXMAN_OP_OVER, texture->image, mask,
+			buffer->image, 0, 0, 0, 0, 0, 0, renderer->width,
+			renderer->height);
+
+	pixman_image_unref(mask);
+
+	return true;
+}
+
+static void pixman_render_quad_with_matrix(struct wlr_renderer *wlr_renderer,
+		const float color[static 4], const float matrix[static 9]) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	struct wlr_pixman_buffer *buffer = renderer->current_buffer;
+
+	struct pixman_color colour = {
+		.red = color[0] * 0xFFFF,
+		.green = color[1] * 0xFFFF,
+		.blue = color[2] * 0xFFFF,
+		.alpha = color[3] * 0xFFFF,
+	};
+
+	pixman_image_t *fill = pixman_image_create_solid_fill(&colour);
+
+	float m[9];
+	memcpy(m, matrix, sizeof(m));
+
+	float width = sqrt(matrix[0] * matrix[0] + matrix[1] * matrix[1]);
+	float height = sqrt(matrix[4] * matrix[4] + matrix[1] * matrix[1]);
```

Comment:
I think this should be `sqrt(matrix[3] * matrix[3] + matrix[4] * matrix[4])`.

Reply:
I was thinking the same thing, but wasn't able to make it crash with `matrix[1]`. I'll go `matrix[3]` just in case.

Reply:
A mix-up won't make it crash, but will give the wrong result if the matrix contains a transformation.

Reply:
Heh I know, all of our examples have square textures, that's why I didn't get any wrong results.